http_client: use HTTP_PROXY environment variable for proxy globally

erain opened the pull request:
- Add http_proxy in global config and set during upstream creation
- HTTP_PROXY url with username and password is also supported with this PR: `http://username:password@host:port`
- Implement `CONNECT` based HTTP support so that HTTPS can work over
  a HTTP proxy. This is important for stackdriver.
- The key is to do proxy negotiation before tls handshake. that's why
  proxy negotiation has to be in upstream.c.

Signed-off-by: Yu Yi <yiyu@google.com>

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

Fixes #2565 

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
https://github.com/fluent/fluent-bit-docs/pull/414
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

erain commented:
### Example configuration for the change

```
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug

[INPUT]
    Name dummy

[OUTPUT]
    Name stackdriver
    Match *
    resource global

[OUTPUT]
    Name http
    Match *
    Host 127.0.0.1
    Port 8080
    Uri /post
```

erain commented:
### Debug log output from testing the change

```
Fluent Bit v1.7.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/10/16 19:32:21] [ info] Configuration:
[2020/10/16 19:32:21] [ info]  flush time     | 1.000000 seconds
[2020/10/16 19:32:21] [ info]  grace          | 5 seconds
[2020/10/16 19:32:21] [ info]  daemon         | 0
[2020/10/16 19:32:21] [ info] ___________
[2020/10/16 19:32:21] [ info]  inputs:
[2020/10/16 19:32:21] [ info]      dummy
[2020/10/16 19:32:21] [ info] ___________
[2020/10/16 19:32:21] [ info]  filters:
[2020/10/16 19:32:21] [ info] ___________
[2020/10/16 19:32:21] [ info]  outputs:
[2020/10/16 19:32:21] [ info]      stackdriver.0
[2020/10/16 19:32:21] [ info]      http.1
[2020/10/16 19:32:21] [ info] ___________
[2020/10/16 19:32:21] [ info]  collectors:
[2020/10/16 19:32:21] [ info] [engine] started (pid=31744)
[2020/10/16 19:32:21] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/10/16 19:32:21] [debug] [storage] [cio stream] new stream registered: dummy.0
[2020/10/16 19:32:21] [ info] [storage] version=1.0.6, initializing...
[2020/10/16 19:32:21] [ info] [storage] in-memory
[2020/10/16 19:32:21] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/10/16 19:32:21] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:32:21] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:32:21] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:32:21] [debug] [output:stackdriver:stackdriver.0] JWT signature:
eyJhbGciOiAiUlMyNTYiLCAidHlwIjogIkpXVCJ9.eyJpc3MiOiAib255eC1vYnNlcnZhYmlsaXR5LXlpeXVAeWl5dS1na2UtZGV2LmlhbS5nc2VydmljZWFjY291bnQuY29tIiwgInNjb3BlIjogImh0dHBzOi8vd3d3Lmdvb2dsZWFwaXMuY29tL2F1dGgvbG9nZ2luZy53cml0ZSIsICJhdWQiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vb2F1dGgyL3Y0L3Rva2VuIiwgImV4cCI6IDE2MDI4Nzk3NDEsICJpYXQiOiAxNjAyODc2NzQxfQ.nBhqmlhp4zVIJ-6MwL_BcOGAKcNjLW9IXZSv1droforg-AvxTokH8rXi0B4vFOLlGqWfsgY6OlHTaqm2jU6Mr9um7ar1y_w2fNIsD-Vb0PYsgEmAbaZICp0m02mIuT7FJ-PkXeL9LRgVZx4WYTVodVYQB0g_57Pd1Kd_lV51fJaNB3GeWcUB7q3ilSKhUA5olHq7pC4rVBucd6ip4sJucbtamlDfBpO72nSiNA93eoXnbWGoyshEahnFqNajNTDs1zLj7I0iKk6hUn_SEEtf4g63TpeKJ5lqaAZczyTjEXMsHtKNlK8BGwaQkEA5HuAX71ugYFkfzScCc6gEPihNIg
[2020/10/16 19:32:21] [debug] flb_establish_proxy
[2020/10/16 19:32:21] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:32:21] [debug] HTTP header=CONNECT www.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: www.googleapis.com
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:32:21] [debug] [upstream] proxy returned 200
[2020/10/16 19:32:21] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:32:21] [debug] HTTP header=POST /oauth2/v4/token HTTP/1.1
Host: www.googleapis.com
Content-Length: 743
Content-Type: application/x-www-form-urlencoded


[2020/10/16 19:32:21] [ info] [oauth2] HTTP Status=200
[2020/10/16 19:32:21] [debug] [oauth2] payload:
{"access_token":"ya29.c.Ko8B4Qdh6Hxh4WwvhHQ3l6A68pk7eZPVBy2nIMx5dVbeyg2cErHuUWeqBv93tTDlMO86H9IJl0RBsXRVLePEw3SxG5DMjn6D7TKHHBwONPY9OzMC_PjJwTQhcNyJGTyc5NAs6KmuA1BWjh5jGUHkX6OvHQWQsDCYqB86-1z4RSUUnUGgTh5apwS1nuwp1atJ1ug","expires_in":3599,"token_type":"Bearer"}
[2020/10/16 19:32:21] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/10/16 19:32:21] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:32:21] [debug] [router] match rule dummy.0:stackdriver.0
[2020/10/16 19:32:21] [debug] [router] match rule dummy.0:http.1
[2020/10/16 19:32:21] [ info] [sp] stream processor started
[2020/10/16 19:32:23] [debug] [task] created task=0x7f14ac144a40 id=0 OK
[2020/10/16 19:32:23] [debug] flb_establish_proxy
[2020/10/16 19:32:23] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:32:23] [debug] HTTP header=CONNECT logging.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: logging.googleapis.com
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:32:23] [debug] [upstream] proxy returned 200
[2020/10/16 19:32:23] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:32:23] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:32:23] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4Qdh6Hxh4WwvhHQ3l6A68pk7eZPVBy2nIMx5dVbeyg2cErHuUWeqBv93tTDlMO86H9IJl0RBsXRVLePEw3SxG5DMjn6D7TKHHBwONPY9OzMC_PjJwTQhcNyJGTyc5NAs6KmuA1BWjh5jGUHkX6OvHQWQsDCYqB86-1z4RSUUnUGgTh5apwS1nuwp1atJ1ug


[2020/10/16 19:32:23] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:32:23] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:32:23 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:32:23] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:32:23] [debug] flb_establish_proxy
[2020/10/16 19:32:23] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:32:23] [debug] HTTP header=CONNECT 127.0.0.1:8080 HTTP/1.1
Proxy-Connection: KeepAlive
Host: 127.0.0.1:8080
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:32:23] [debug] [upstream] proxy returned 200
[2020/10/16 19:32:23] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:32:23] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:32:23] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:32:23] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n1Rh2WKhSBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}
```

erain commented:
### Valgrind output that shows no leaks or memory corruption was found

```
==32119== Memcheck, a memory error detector
==32119== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==32119== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==32119== Command: ./bin/fluent-bit -c fluent-bit.config
==32119==
Fluent Bit v1.7.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/10/16 19:36:00] [ info] Configuration:
[2020/10/16 19:36:00] [ info]  flush time     | 1.000000 seconds
[2020/10/16 19:36:00] [ info]  grace          | 5 seconds
[2020/10/16 19:36:00] [ info]  daemon         | 0
[2020/10/16 19:36:00] [ info] ___________
[2020/10/16 19:36:00] [ info]  inputs:
[2020/10/16 19:36:00] [ info]      dummy
[2020/10/16 19:36:00] [ info] ___________
[2020/10/16 19:36:00] [ info]  filters:
[2020/10/16 19:36:00] [ info] ___________
[2020/10/16 19:36:00] [ info]  outputs:
[2020/10/16 19:36:00] [ info]      stackdriver.0
[2020/10/16 19:36:00] [ info]      http.1
[2020/10/16 19:36:00] [ info] ___________
[2020/10/16 19:36:00] [ info]  collectors:
[2020/10/16 19:36:00] [ info] [engine] started (pid=32119)
[2020/10/16 19:36:00] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/10/16 19:36:00] [debug] [storage] [cio stream] new stream registered: dummy.0
[2020/10/16 19:36:00] [ info] [storage] version=1.0.6, initializing...
[2020/10/16 19:36:00] [ info] [storage] in-memory
[2020/10/16 19:36:00] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/10/16 19:36:01] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:36:01] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:36:01] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:36:01] [debug] [output:stackdriver:stackdriver.0] JWT signature:
eyJhbGciOiAiUlMyNTYiLCAidHlwIjogIkpXVCJ9.eyJpc3MiOiAib255eC1vYnNlcnZhYmlsaXR5LXlpeXVAeWl5dS1na2UtZGV2LmlhbS5nc2VydmljZWFjY291bnQuY29tIiwgInNjb3BlIjogImh0dHBzOi8vd3d3Lmdvb2dsZWFwaXMuY29tL2F1dGgvbG9nZ2luZy53cml0ZSIsICJhdWQiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vb2F1dGgyL3Y0L3Rva2VuIiwgImV4cCI6IDE2MDI4Nzk5NjEsICJpYXQiOiAxNjAyODc2OTYxfQ.UMlC6b3VCb__6C43QqDmSWTiOfKHdeZeHQQ8MkbiuSpIkOVSa6vCXiCrOv_7tBMCmI_2c31qBMh_nPJ_qgCSLpmrd6fsQqgvjhrjYd7Gw2gjUvmrAoyfH8g4X6D0XbFgLPjyAikWhJ__LKG6jFzjTE2IbnrTRP01Uu13Q8XSwEJ4i0XRDrSyhwRs3erJw5H7-Gpm3Op_6-0rjy5hfjYPykibhW_-SicbxQ6VftYiydimplSJZUEkNxUf433iwQkDHaSrBdN2S968fKX_6ZSLy-QVryS-_psWIezq_G3kBuKSRRJ36MuOjk3ZFwMFi0Ix2vi3yxfeurpcxdjxYttLvg
[2020/10/16 19:36:01] [debug] flb_establish_proxy
[2020/10/16 19:36:01] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:36:01] [debug] HTTP header=CONNECT www.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: www.googleapis.com
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:36:01] [debug] [upstream] proxy returned 200
[2020/10/16 19:36:02] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:02] [debug] HTTP header=POST /oauth2/v4/token HTTP/1.1
Host: www.googleapis.com
Content-Length: 743
Content-Type: application/x-www-form-urlencoded


[2020/10/16 19:36:02] [ info] [oauth2] HTTP Status=200
[2020/10/16 19:36:02] [debug] [oauth2] payload:
{"access_token":"ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4","expires_in":3599,"token_type":"Bearer"}
[2020/10/16 19:36:02] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/10/16 19:36:02] [debug] config->http_proxy: http://localhost:8888
[2020/10/16 19:36:02] [debug] [router] match rule dummy.0:stackdriver.0
[2020/10/16 19:36:02] [debug] [router] match rule dummy.0:http.1
[2020/10/16 19:36:02] [ info] [sp] stream processor started
==32119== Warning: client switching stacks?  SP change: 0x6805948 --> 0x792e0a0
==32119==          to suppress, use: --max-stackframe=17991512 or greater
==32119== Warning: client switching stacks?  SP change: 0x792e018 --> 0x6805948
==32119==          to suppress, use: --max-stackframe=17991376 or greater
==32119== Warning: client switching stacks?  SP change: 0x6805948 --> 0x792e018
==32119==          to suppress, use: --max-stackframe=17991376 or greater
==32119==          further instances of this message will not be shown.
[2020/10/16 19:36:04] [debug] [task] created task=0x7927c10 id=0 OK
[2020/10/16 19:36:04] [debug] flb_establish_proxy
[2020/10/16 19:36:04] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:36:04] [debug] HTTP header=CONNECT logging.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: logging.googleapis.com
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:36:04] [debug] [upstream] proxy returned 200
[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:36:05] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:05] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4


[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:36:05 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:36:05] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:36:05] [debug] flb_establish_proxy
[2020/10/16 19:36:05] [debug] [PROXY] using HTTP CONNECT for proxy
[2020/10/16 19:36:05] [debug] HTTP header=CONNECT 127.0.0.1:8080 HTTP/1.1
Proxy-Connection: KeepAlive
Host: 127.0.0.1:8080
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/16 19:36:05] [debug] [upstream] proxy returned 200
[2020/10/16 19:36:05] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:05] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:36:05] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:36:05] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n2Ix5OM5OBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/16 19:36:05] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/16 19:36:05] [debug] [task] destroy task=0x7927c10 (task_id=0)
[2020/10/16 19:36:05] [debug] [task] created task=0x7d125b0 id=0 OK
[2020/10/16 19:36:05] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:36:05] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:05] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4


[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:36:05] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:36:05 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:36:05] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:36:05] [debug] [upstream] KA connection #27 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:05] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:05] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:36:05] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:36:05] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n2JRe1ur+Bp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/16 19:36:05] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/16 19:36:05] [debug] [task] destroy task=0x7d125b0 (task_id=0)
[2020/10/16 19:36:06] [debug] [task] created task=0x7d79e80 id=0 OK
[2020/10/16 19:36:06] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:06] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:36:06] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:06] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4


[2020/10/16 19:36:06] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:36:06] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:36:06 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:36:06] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:36:06] [debug] [upstream] KA connection #27 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:06] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:06] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:36:06] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:36:06] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n2JSJeqvuBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/16 19:36:06] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/16 19:36:06] [debug] [task] destroy task=0x7d79e80 (task_id=0)
[2020/10/16 19:36:07] [debug] [task] created task=0x7de1750 id=0 OK
[2020/10/16 19:36:07] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:07] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:36:07] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:07] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4


[2020/10/16 19:36:07] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:36:07] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:36:07 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:36:07] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:36:07] [debug] [upstream] KA connection #27 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:07] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:07] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:36:07] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:36:07] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n2JiJ2WFCBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/16 19:36:07] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/16 19:36:07] [debug] [task] destroy task=0x7de1750 (task_id=0)
^C[2020/10/16 19:36:08] [engine] caught signal (SIGINT)
[2020/10/16 19:36:08] [debug] [task] created task=0x7e09030 id=0 OK
[2020/10/16 19:36:08] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:08] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/16 19:36:08] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:08] [debug] HTTP header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Ko8B4QeYXqizujR0012veiAvwooKBVsM4dQ8rWCJtuySPn6lgVxiQnfH-304FgXWxaYCYAAnMtI_ZDAjsrcHc3urXBd3LTSbMSvLDd1q78r7WzRxYkYC4sexFzgns3wpfyQRoS-98BjEF4TSMpGiz1pmjLcudDoF-kLoRfX0s-83AyWtOEmCWrou2Z1cOuoFnB4


[2020/10/16 19:36:08] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/16 19:36:08] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 16 Oct 2020 19:36:08 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/16 19:36:08] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/16 19:36:08] [debug] [upstream] KA connection #27 to localhost:8888 has been assigned (recycled)
[2020/10/16 19:36:08] [debug] [PROXY] not using http_proxy for header
[2020/10/16 19:36:08] [debug] [HTTP_OUT] http_client proxy host: (null) port: 0
[2020/10/16 19:36:08] [debug] HTTP header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/16 19:36:08] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX4n2JyK5uamBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/16 19:36:08] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/16 19:36:08] [ warn] [engine] service will stop in 5 seconds
[2020/10/16 19:36:08] [debug] [task] destroy task=0x7e09030 (task_id=0)
[2020/10/16 19:36:08] [debug] [input chunk] dummy.0 is paused, cannot append records
^C[2020/10/16 19:36:09] [engine] caught signal (SIGINT)
==32119==
==32119== HEAP SUMMARY:
==32119==     in use at exit: 2,262,649 bytes in 8,503 blocks
==32119==   total heap usage: 83,563 allocs, 75,060 frees, 11,708,973 bytes allocated
==32119==
==32119== LEAK SUMMARY:
==32119==    definitely lost: 0 bytes in 0 blocks
==32119==    indirectly lost: 0 bytes in 0 blocks
==32119==      possibly lost: 2,215,221 bytes in 8,454 blocks
==32119==    still reachable: 47,428 bytes in 49 blocks
==32119==         suppressed: 0 bytes in 0 blocks
==32119== Rerun with --leak-check=full to see details of leaked memory
==32119==
==32119== For counts of detected and suppressed errors, rerun with: -v
==32119== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

erain commented:
I will also create a documentation PR during the review of this PR.

erain commented:
cc @liuyuan10 @courageJ

liuyuan10 commented on src/flb_config.c:
```diff
@@ -161,6 +161,8 @@ struct flb_config *flb_config_init()
     config->http_port    = flb_strdup(FLB_CONFIG_HTTP_PORT);
 #endif
 
+    config->http_proxy = getenv("HTTP_PROXY");
```
usually when HTTP_PROXY="" (empty), proxy should be disabled. could you set it to NULL when it's empty string?

erain replied:
It looks like `getenv` already returns `NULL` in this case:

```
/* Return the value of envariable NAME, or NULL if it doesn't exist.  */
extern char *getenv (const char *__name) __THROW __nonnull ((1)) __wur;
```

erain replied:
Oh I see, you want to prevent the case where `HTTP_PROXY=""`. OK, fixing that now.

erain replied:
Done.

liuyuan10 replied:
nit: could you squash the commits? having separete commits for small fixes is not clean.

erain replied:
Sound good. (I did squash before in github but the previous review got lost). At this point it should not matter.

edsiper commented on src/flb_config.c:
```diff
@@ -161,6 +161,12 @@ struct flb_config *flb_config_init()
 #endif
 
+    config->http_proxy = getenv("HTTP_PROXY");
+    if (strcmp(config->http_proxy, "")) {
```
there is a chance that ```config->http_proxy``` becomes NULL and the strcmp will crash:

```
RETURN VALUE
       The getenv() function returns a pointer to the value in the environment, or NULL if there is no match.
```

in the same conditional make sure config->http_proxy is not NULL

edsiper commented on src/flb_http_client.c:
```diff
@@ -614,19 +632,30 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
     /* FIXME: handler for HTTPS proxy */
-    if (!proxy) {
+    if (proxy) {
+        flb_debug("[PROXY] using http_proxy %s for header", proxy);
```
lower case prefix ```[PROXY]``` -> ```[http_client] proxy: ```

edsiper commented on src/flb_http_client.c:
```diff
@@ -614,19 +632,30 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
     }
-    else {
+    else if (method == FLB_HTTP_CONNECT) {
+        flb_debug("[PROXY] using HTTP CONNECT for proxy");
```
lower case prefix [PROXY] -> [http_client] proxy:

edsiper commented on src/flb_http_client.c:
```diff
@@ -614,19 +632,30 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
+                       flags & FLB_HTTP_10 ? 0 : 1);
+    }
+    else {
+        flb_debug("[PROXY] not using http_proxy for header");
```
lower case prefix [PROXY] -> [http_client] proxy:

edsiper commented on src/flb_http_client.c:
```diff
@@ -684,8 +713,10 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
 
     /* Check proxy data */
     if (proxy) {
+        flb_debug("[PROXY] Using http_proxy: %s", proxy);
```
lower case prefix [PROXY] -> [http_client] proxy:

edsiper commented on src/flb_http_client.c:
```diff
@@ -684,8 +713,10 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
+        flb_debug("[PROXY] Using http_proxy: %s", proxy);
         ret = proxy_parse(proxy, c);
         if (ret != 0) {
+            flb_debug("Something wrong with the http_proxy parsing");
```
lower case prefix [PROXY] -> [http_client] proxy:

edsiper commented on src/flb_http_client.c:
```diff
@@ -1066,6 +1097,7 @@ int flb_http_do(struct flb_http_client *c, size_t *bytes)
     }
 #endif
 
+    flb_debug("HTTP header=%s", c->header_buf);
```
lower case prefix  ```[http_client] ...```

edsiper commented on src/flb_io.c:
```diff
@@ -63,6 +63,7 @@
 #include <fluent-bit/flb_network.h>
 #include <fluent-bit/flb_engine.h>
 #include <fluent-bit/flb_thread.h>
+#include <fluent-bit/flb_http_client.h>
```
the ```io``` interface should not be aware about HTTP contexts. Can you move the new functions to the flb_http_client.c ?

note: make sure to prefix the functions with the proper interface, e.g: ```flb_http_client_proxy_connect```

edsiper reviewed (changes requested):
thanks for this!

please address the comments then I can continue the final review.

edsiper commented:
FYI:

in the valgrind output there are memory leaks:

```
==32119== HEAP SUMMARY:
==32119==     in use at exit: 2,262,649 bytes in 8,503 blocks
==32119==   total heap usage: 83,563 allocs, 75,060 frees, 11,708,973 bytes allocated
```

erain commented:
Thanks for the review! I addressed all the comments and currently doing another round of testing. Should be able to push a new commit today. 

Also, cc @rsgowman for reviewing the possible memory leaks.

erain commented:
RE - "memory leak": thanks a lot for @rsgowman identifying that there were no real memory leaks

> "You seem to have hit ^C while it was shutting down. When testing fluentbit; I hit ^C twice to stop it. The first tells it to shutdown, but it takes 5s to do so, so I usually just hit ^C again. However, that second ^C interrupts the cleanup. Try just waiting."

Below is the valgrind run from the latest commit (which I am going to push in a few mins) and it shows no memory leaks.



```
$ valgrind ./bin/fluent-bit -c fluent-bit.config
==23843== Memcheck, a memory error detector
==23843== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==23843== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==23843== Command: ./bin/fluent-bit -c fluent-bit.config
==23843==
Fluent Bit v1.7.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/10/23 19:39:28] [ info] Configuration:
[2020/10/23 19:39:28] [ info]  flush time     | 1.000000 seconds
[2020/10/23 19:39:28] [ info]  grace          | 5 seconds
[2020/10/23 19:39:28] [ info]  daemon         | 0
[2020/10/23 19:39:28] [ info] ___________
[2020/10/23 19:39:28] [ info]  inputs:
[2020/10/23 19:39:28] [ info]      dummy
[2020/10/23 19:39:28] [ info] ___________
[2020/10/23 19:39:28] [ info]  filters:
[2020/10/23 19:39:28] [ info] ___________
[2020/10/23 19:39:28] [ info]  outputs:
[2020/10/23 19:39:28] [ info]      stackdriver.0
[2020/10/23 19:39:28] [ info]      http.1
[2020/10/23 19:39:28] [ info] ___________
[2020/10/23 19:39:28] [ info]  collectors:
[2020/10/23 19:39:28] [ info] [engine] started (pid=23843)
[2020/10/23 19:39:28] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/10/23 19:39:28] [debug] [storage] [cio stream] new stream registered: dummy.0
[2020/10/23 19:39:28] [ info] [storage] version=1.0.6, initializing...
[2020/10/23 19:39:28] [ info] [storage] in-memory
[2020/10/23 19:39:28] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/10/23 19:39:29] [debug] config->http_proxy: http://localhost:8888
[2020/10/23 19:39:29] [debug] config->http_proxy: http://localhost:8888
[2020/10/23 19:39:29] [debug] config->http_proxy: http://localhost:8888
[2020/10/23 19:39:29] [debug] [output:stackdriver:stackdriver.0] JWT signature:
eyJhbGciOiAiUlMyNTYiLCAidHlwIjogIkpXVCJ9.eyJpc3MiOiAib255eC1vYnNlcnZhYmlsaXR5LXlpeXVAeWl5dS1na2UtZGV2LmlhbS5nc2VydmljZWFjY291bnQuY29tIiwgInNjb3BlIjogImh0dHBzOi8vd3d3Lmdvb2dsZWFwaXMuY29tL2F1dGgvbG9nZ2luZy53cml0ZSIsICJhdWQiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vb2F1dGgyL3Y0L3Rva2VuIiwgImV4cCI6IDE2MDM0ODQ5NjksICJpYXQiOiAxNjAzNDgxOTY5fQ.xld7YpOJ8IXgEWu-JPLqFTmMbehBrKQM_GtU1SXndn5fHgtp5Ss9LuAak2GBSVVoi5_fanLYdDZfM2Fl86ak3vvFnDXBBpoS0tlzNgxqX3Q_DsO9ivkJvbtqPj5ih90iRPFy_4-YNS8D0xIdOw4jb3IkCEi_0Bxf8idIUIa6SOUjgcSu0rvsAWmO3cGvqQejZKPc8gARVTsMzgPMENt6tvO0lndoSqOlQ-d9379DNTZfLf0fHTxdNjB5cm03GKucF5gJuc5sciYefaVh-h3mdGMbIJMm2DkAVMIehMp78H1DYQd4vECwxqRZAVPKJV47UdWPrQgVvONPHSCql64bXg
[2020/10/23 19:39:29] [debug] [http_client] using HTTP CONNECT for proxy
[2020/10/23 19:39:29] [debug] [http_client] header=CONNECT www.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: www.googleapis.com
Content-Length: 0
User-Agent: Fluent-Bit


[2020/10/23 19:39:29] [debug] [upstream] proxy returned 200
[2020/10/23 19:39:29] [debug] [http_client] flb_http_client_proxy_connect connection #20 connected to localhost:8888.
[2020/10/23 19:39:30] [debug] [http_client] not using http_proxy for header

... 

[2020/10/23 19:39:40] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/23 19:39:40] [debug] [task] destroy task=0x7fecab0 (task_id=0)
[2020/10/23 19:39:41] [debug] [task] created task=0x8054390 id=0 OK
[2020/10/23 19:39:41] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/10/23 19:39:41] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/10/23 19:39:41] [debug] [http_client] not using http_proxy for header
[2020/10/23 19:39:41] [debug] [http_client] header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Kp0B4geHwYIlNESOeaT3xg38IFwTT7Dl5MWZcpM6cYC04wZi20GPrkh8lakL6TornEhNE42VaBAdbe8K4k6WzrOx2jtJozeiMQW9E11guEhoViDx_-uetDWNdAbGzIFH6omdLIEG4PAwOlGDJULiapj2A4hwpxvlEHfDoJjWEgsbE18McGDNXGGF0M4ShbxNVZJgJnd2WJz8PfgyMygJxQ


[2020/10/23 19:39:41] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/10/23 19:39:41] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Fri, 23 Oct 2020 19:39:41 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/10/23 19:39:41] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/10/23 19:39:41] [debug] [upstream] KA connection #27 to localhost:8888 has been assigned (recycled)
[2020/10/23 19:39:41] [debug] [http_client] not using http_proxy for header
[2020/10/23 19:39:41] [debug] [http_client] proxy host: (null) port: 0
[2020/10/23 19:39:41] [debug] [http_client] header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/10/23 19:39:41] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX5MxfR2WyqCBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/10/23 19:39:41] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/10/23 19:39:41] [debug] [task] destroy task=0x8054390 (task_id=0)
^C[2020/10/23 19:39:42] [engine] caught signal (SIGINT)
[2020/10/23 19:39:42] [ warn] [engine] service will stop in 5 seconds
[2020/10/23 19:39:42] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/10/23 19:39:43] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/10/23 19:39:43] [debug] [upstream] KA connection #27 to localhost:8888 has been disconnected by the remote service
[2020/10/23 19:39:44] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/10/23 19:39:45] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/10/23 19:39:46] [ info] [engine] service stopped
==23843==
==23843== HEAP SUMMARY:
==23843==     in use at exit: 0 bytes in 0 blocks
==23843==   total heap usage: 83,474 allocs, 83,474 frees, 14,004,049 bytes allocated
==23843==
==23843== All heap blocks were freed -- no leaks are possible
==23843==
==23843== For counts of detected and suppressed errors, rerun with: -v
==23843== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

```

erain reviewed (commented):
All comments addressed. This PR is ready for another round of review.

erain commented:
Friendly ping :D @edsiper : could you help with another round of review? Many thanks!

edsiper reviewed (changes requested):
Please review my comments, not all of them might need a change, but just an extra check

erain commented:
Hi @edsiper I've addressed and revolved all comments. Also the "potential memory leaks" was explained and tested again in https://github.com/fluent/fluent-bit/pull/2694#issuecomment-715554719 . 

Could you help point out what might I miss now?

edsiper commented:
@erain did you force-push the changes ? Looking at the changes it was pretty much the same the last time...

erain commented:
Hi @edsiper , yes I did force push for commit (so all the comments are marked as (outdated)). Maybe that is the wrong way for addressing comments, sorry about that and will remember not doing that the next time. 

Would you mind to see the updated code one more time in "File Changed" tab? (again sorry for the trouble caused)

erain commented:
I just pushed a commit which added the username/password support in HTTP_PROXY url. 

Also I added that to this PR's description: 

- HTTP_PROXY url with username and password is also supported with this PR: `http://username:password@host:port`

erain commented:
I tested the code change as in https://github.com/fluent/fluent-bit/pull/2694#issuecomment-710484172 to show both `http` and `stackdriver` output plugins worked, with `BasicAuth foo bar` in tinyproxy's configuration and `HTTP_PROXY=http://foo:bar@localhost:8888`

Also I ran `valgrind` again to ensure there were no memory leaks.

```

# yiyu @ yiyu-fluent-bit-dev in ~/src/fluent-bit/build on git:feature/proxy_url_parsing o [19:37:48]
$ valgrind ./bin/fluent-bit -c fluent-bit.config
==31452== Memcheck, a memory error detector
==31452== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==31452== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==31452== Command: ./bin/fluent-bit -c fluent-bit.config
==31452==
Fluent Bit v1.7.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/11/05 19:37:51] [ info] Configuration:
[2020/11/05 19:37:51] [ info]  flush time     | 1.000000 seconds
[2020/11/05 19:37:51] [ info]  grace          | 5 seconds
[2020/11/05 19:37:51] [ info]  daemon         | 0
[2020/11/05 19:37:51] [ info] ___________
[2020/11/05 19:37:51] [ info]  inputs:
[2020/11/05 19:37:51] [ info]      dummy
[2020/11/05 19:37:51] [ info] ___________
[2020/11/05 19:37:51] [ info]  filters:
[2020/11/05 19:37:51] [ info] ___________
[2020/11/05 19:37:51] [ info]  outputs:
[2020/11/05 19:37:51] [ info]      stackdriver.0
[2020/11/05 19:37:51] [ info]      http.1
[2020/11/05 19:37:51] [ info] ___________
[2020/11/05 19:37:51] [ info]  collectors:
[2020/11/05 19:37:51] [ info] [engine] started (pid=31452)
[2020/11/05 19:37:51] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/11/05 19:37:51] [debug] [storage] [cio stream] new stream registered: dummy.0
[2020/11/05 19:37:51] [ info] [storage] version=1.0.6, initializing...
[2020/11/05 19:37:51] [ info] [storage] in-memory
[2020/11/05 19:37:51] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/11/05 19:37:52] [debug] [upstream] config->http_proxy: http://foo:bar@localhost:8888
[2020/11/05 19:37:52] [debug] [upstream] parsed http_proxy: protocol http, host localhost port 8888 username foo password bar
[2020/11/05 19:37:52] [debug] [upstream] config->http_proxy: http://foo:bar@localhost:8888
[2020/11/05 19:37:52] [debug] [upstream] parsed http_proxy: protocol http, host localhost port 8888 username foo password bar
[2020/11/05 19:37:52] [debug] [upstream] config->http_proxy: http://foo:bar@localhost:8888
[2020/11/05 19:37:52] [debug] [upstream] parsed http_proxy: protocol http, host localhost port 8888 username foo password bar
[2020/11/05 19:37:52] [debug] [output:stackdriver:stackdriver.0] JWT signature:
eyJhbGciOiAiUlMyNTYiLCAidHlwIjogIkpXVCJ9.eyJpc3MiOiAib255eC1vYnNlcnZhYmlsaXR5LXlpeXVAeWl5dS1na2UtZGV2LmlhbS5nc2VydmljZWFjY291bnQuY29tIiwgInNjb3BlIjogImh0dHBzOi8vd3d3Lmdvb2dsZWFwaXMuY29tL2F1dGgvbG9nZ2luZy53cml0ZSIsICJhdWQiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vb2F1dGgyL3Y0L3Rva2VuIiwgImV4cCI6IDE2MDQ2MDgwNzIsICJpYXQiOiAxNjA0NjA1MDcyfQ.KacU21x7hJCUDqnLzrzyS5GkuTQ_WmiAs6n5r0BJXAZE4rgekxca6C9J1rtGZIUs8kyJQ-PEaGTWuCn8Y0YyQW3zT-Svqzuncop5VWxZrZBao6h7TNN5AYXDv6kll2mHTuHyywYJP5kWfnQvNhuFRIHL5cdSX15z5EnIUuj_ygWCnlpraTKgtmDk2kWwaqwPQhSCefzQuDVc5M1pcFN4AC7t3wvmkFluMTFeEpFoGsdui3d3-y0QkVu994HIQz0gzUgRfIXmnHoU-0Jsv1QzWV-c6cfQikQyGwp0g0HHIOolJlaNyc0rKQwxPlPd-rLJkGwAs91_2x1GC-aUzLgOgA
[2020/11/05 19:37:53] [debug] [upstream] establishing http tunneling to proxy: host localhost port 8888
[2020/11/05 19:37:53] [debug] [http_client] using HTTP CONNECT for proxy: proxy host www.googleapis.com, proxy port 443
[2020/11/05 19:37:53] [debug] [upstream] proxy uses username foo password bar
[2020/11/05 19:37:53] [debug] [http_client] header=CONNECT www.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: www.googleapis.com
Content-Length: 0
Proxy-Authorization: Basic Zm9vOmJhcg==
User-Agent: Fluent-Bit


[2020/11/05 19:37:53] [debug] [upstream] proxy returned 200
[2020/11/05 19:37:53] [debug] [http_client] flb_http_client_proxy_connect connection #20 connected to localhost:8888.
[2020/11/05 19:37:54] [debug] [http_client] not using http_proxy for header
[2020/11/05 19:37:54] [debug] [http_client] header=POST /oauth2/v4/token HTTP/1.1
Host: www.googleapis.com
Content-Length: 743
Content-Type: application/x-www-form-urlencoded


[2020/11/05 19:37:54] [ info] [oauth2] HTTP Status=200
[2020/11/05 19:37:54] [debug] [oauth2] payload:
{"access_token":"ya29.c.Kp0B5Achk3sAIFllADtjYxO53dGqN537MOBHyPD3EaON_emJuxzuyHD_NkU9tU9luSl4S7h1dfjhL_wvP7SyV95YUaCGA5IMnDK5FJ7DRuGwAPtX9LdWj2YkCw8WW2gujOVgwQrINOEGiEzCrPMUO-gsPPm1B3dY7tfiU49NhU7oI3Js9fLbvr011somatX_F1sXfQ5GcC4zE4kjMwqn7g","expires_in":3599,"token_type":"Bearer"}
[2020/11/05 19:37:54] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/11/05 19:37:54] [debug] [upstream] config->http_proxy: http://foo:bar@localhost:8888
[2020/11/05 19:37:54] [debug] [upstream] parsed http_proxy: protocol http, host localhost port 8888 username foo password bar
[2020/11/05 19:37:54] [debug] [router] match rule dummy.0:stackdriver.0
[2020/11/05 19:37:54] [debug] [router] match rule dummy.0:http.1
[2020/11/05 19:37:54] [ info] [sp] stream processor started
==31452== Warning: client switching stacks?  SP change: 0x6805948 --> 0x79ba510
==31452==          to suppress, use: --max-stackframe=18566088 or greater
==31452== Warning: client switching stacks?  SP change: 0x79ba488 --> 0x6805948
==31452==          to suppress, use: --max-stackframe=18565952 or greater
==31452== Warning: client switching stacks?  SP change: 0x6805948 --> 0x79ba488
==31452==          to suppress, use: --max-stackframe=18565952 or greater
==31452==          further instances of this message will not be shown.
[2020/11/05 19:37:55] [debug] [task] created task=0x79b4080 id=0 OK
[2020/11/05 19:37:55] [debug] [upstream] establishing http tunneling to proxy: host localhost port 8888
[2020/11/05 19:37:55] [debug] [http_client] using HTTP CONNECT for proxy: proxy host logging.googleapis.com, proxy port 443
[2020/11/05 19:37:55] [debug] [upstream] proxy uses username foo password bar
[2020/11/05 19:37:55] [debug] [http_client] header=CONNECT logging.googleapis.com:443 HTTP/1.1
Proxy-Connection: KeepAlive
Host: logging.googleapis.com
Content-Length: 0
Proxy-Authorization: Basic Zm9vOmJhcg==
User-Agent: Fluent-Bit


[2020/11/05 19:37:55] [debug] [upstream] proxy returned 200
[2020/11/05 19:37:55] [debug] [http_client] flb_http_client_proxy_connect connection #26 connected to localhost:8888.
[2020/11/05 19:37:56] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/11/05 19:37:56] [debug] [http_client] not using http_proxy for header
[2020/11/05 19:37:56] [debug] [http_client] header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Kp0B5Achk3sAIFllADtjYxO53dGqN537MOBHyPD3EaON_emJuxzuyHD_NkU9tU9luSl4S7h1dfjhL_wvP7SyV95YUaCGA5IMnDK5FJ7DRuGwAPtX9LdWj2YkCw8WW2gujOVgwQrINOEGiEzCrPMUO-gsPPm1B3dY7tfiU49NhU7oI3Js9fLbvr011somatX_F1sXfQ5GcC4zE4kjMwqn7g


[2020/11/05 19:37:56] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/11/05 19:37:56] [debug] [output:stackdriver:stackdriver.0] HTTP data=HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Vary: X-Origin
Vary: Referer
Date: Thu, 05 Nov 2020 19:37:56 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Accept-Ranges: none
Vary: Origin,Accept-Encoding
Transfer-Encoding: chunked

{}

[2020/11/05 19:37:56] [debug] [upstream] KA connection #26 to localhost:8888 is now available
[2020/11/05 19:37:56] [debug] [upstream] establishing http tunneling to proxy: host localhost port 8888
[2020/11/05 19:37:56] [debug] [http_client] using HTTP CONNECT for proxy: proxy host 127.0.0.1, proxy port 8080
[2020/11/05 19:37:56] [debug] [upstream] proxy uses username foo password bar
[2020/11/05 19:37:56] [debug] [http_client] header=CONNECT 127.0.0.1:8080 HTTP/1.1
Proxy-Connection: KeepAlive
Host: 127.0.0.1:8080
Content-Length: 0
Proxy-Authorization: Basic Zm9vOmJhcg==
User-Agent: Fluent-Bit


[2020/11/05 19:37:56] [debug] [upstream] proxy returned 200
[2020/11/05 19:37:56] [debug] [http_client] flb_http_client_proxy_connect connection #27 connected to localhost:8888.
[2020/11/05 19:37:56] [debug] [http_client] not using http_proxy for header
[2020/11/05 19:37:56] [debug] [http_client] proxy host: (null) port: 0
[2020/11/05 19:37:56] [debug] [http_client] header=POST /post HTTP/1.1
Host: 127.0.0.1:8080
Content-Length: 26
Content-Type: application/msgpack
User-Agent: Fluent-Bit


[2020/11/05 19:37:56] [ info] [output:http:http.1] 127.0.0.1:8080, HTTP status=200
{
  "args": {},
  "data": "data:application/octet-stream;base64,ktcAX6RUkh5NSkSBp21lc3NhZ2WlZHVtbXk=",
  "files": {},
  "form": {},
  "headers": {
    "Content-Length": "26",
    "Content-Type": "application/msgpack",
    "Host": "127.0.0.1:8080",
    "User-Agent": "Fluent-Bit"
  },
  "json": null,
  "origin": "172.17.0.1",
  "url": "http://127.0.0.1:8080/post"
}

[2020/11/05 19:37:56] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/11/05 19:37:56] [debug] [task] destroy task=0x79b4080 (task_id=0)
[2020/11/05 19:37:56] [debug] [task] created task=0x7d9ab50 id=0 OK
[2020/11/05 19:37:56] [debug] [upstream] KA connection #26 to localhost:8888 has been assigned (recycled)
[2020/11/05 19:37:56] [debug] [output:stackdriver:stackdriver.0] [logging.googleapis.com/monitored_resource] not found in the payload
[2020/11/05 19:37:56] [debug] [http_client] not using http_proxy for header
[2020/11/05 19:37:56] [debug] [http_client] header=POST /v2/entries:write HTTP/1.1
Host: logging.googleapis.com
Content-Length: 209
User-Agent: Fluent-Bit
Content-Type: application/json
Authorization: Bearer ya29.c.Kp0B5Achk3sAIFllADtjYxO53dGqN537MOBHyPD3EaON_emJuxzuyHD_NkU9tU9luSl4S7h1dfjhL_wvP7SyV95YUaCGA5IMnDK5FJ7DRuGwAPtX9LdWj2YkCw8WW2gujOVgwQrINOEGiEzCrPMUO-gsPPm1B3dY7tfiU49NhU7oI3Js9fLbvr011somatX_F1sXfQ5GcC4zE4kjMwqn7g

...

[2020/11/05 19:38:28] [debug] [upstream] KA connection #27 to localhost:8888 is now available
[2020/11/05 19:38:28] [ warn] [engine] service will stop in 5 seconds
[2020/11/05 19:38:28] [debug] [task] destroy task=0x5d92ac0 (task_id=0)
[2020/11/05 19:38:28] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/11/05 19:38:29] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/11/05 19:38:30] [debug] [upstream] KA connection #27 to localhost:8888 has been disconnected by the remote service
[2020/11/05 19:38:30] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/11/05 19:38:31] [debug] [input chunk] dummy.0 is paused, cannot append records
[2020/11/05 19:38:32] [ info] [engine] service stopped
==31452==
==31452== HEAP SUMMARY:
==31452==     in use at exit: 0 bytes in 0 blocks
==31452==   total heap usage: 86,686 allocs, 86,686 frees, 24,499,798 bytes allocated
==31452==
==31452== All heap blocks were freed -- no leaks are possible
==31452==
==31452== For counts of detected and suppressed errors, rerun with: -v
==31452== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

```

edsiper commented on include/fluent-bit/flb_config.h:
```diff
@@ -144,6 +144,16 @@ struct flb_config {
 #endif
 
+    /* There are two ways to use proxy in fluent-bit:
+         #1 Similar with http and datadog plugin, passing proxy directly to
```
multiline comment still not set: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#multiline-comment

erain replied:
Fixed.

edsiper commented on plugins/out_http/http.c:
```diff
@@ -113,6 +113,9 @@ static int http_post(struct flb_out_http *ctx,
                         ctx->proxy, 0);
 
+
+    flb_debug("[http_client] proxy host: %s port: %i", c->proxy.host, c->proxy.port);
```
since this debug message is inside a plugin code, it should use flb_plg_debug(ctx->ins, "..."), on this case don't need the component prefix since the API will put it there automatically

erain replied:
Fixed.

edsiper commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -1862,6 +1862,7 @@ static void cb_stackdriver_flush(const void *data, size_t bytes,
     else {
         /* The request was issued successfully, validate the 'error' field */
         flb_plg_debug(ctx->ins, "HTTP Status=%i", c->resp.status);
+        flb_plg_debug(ctx->ins, "HTTP data=%s", c->resp.data);
```
do not print the payload since it might be corrupted, you can get the data with _debug.http.response_payload configuration property

erain replied:
Fixed by removing this line.

edsiper commented on src/flb_http_client.c:
```diff
@@ -1017,12 +1047,25 @@ int flb_http_basic_auth(struct flb_http_client *c,
-                              FLB_HTTP_HEADER_AUTH,
-                              sizeof(FLB_HTTP_HEADER_AUTH) - 1,
+                              header,
+                              strlen(header),
```
please do strlen() before this function, not as a parameter

erain replied:
Fixed.

edsiper commented on src/flb_io.c:
```diff
@@ -610,7 +623,7 @@ int flb_io_net_write(struct flb_upstream_conn *u_conn, const void *data,
               th, len);
 
-    if (u->flags & FLB_IO_TCP) {
+    if (!u_conn->tls_session) {
```
I am not sure about this change, would you please check the behavior for different plugins that have FLB_IO_TCP and FLB_IO_OPT_TLS ?

liuyuan10 replied:
A bit context on this line of change. Previously it's using the flag to decide if the read/write should be sent through TCP or TLS after encryption. For TCP connections, there is no behavior change because anyway there is no tls_session established (net_io_tls_handshake is only called when FLB_IO_TLS).

For FLB_IO_TLS, this change will not introduce issue because anyway the following code is assuming tls_session is not null. This change simply gives it a chance to send TCP payload without TLS encryption before TLS session is established. It's needed in the proxy case where proxy connect must be sent before TLS session.

edsiper commented on src/flb_upstream.c:
```diff
@@ -84,13 +92,42 @@ struct flb_upstream *flb_upstream_create(struct flb_config *config,
+    /* Set upstream to the http_proxy if it is specified. */
+    if (config->http_proxy) {
+        flb_debug("[upstream] config->http_proxy: %s", config->http_proxy);
+        ret = flb_utils_proxy_url_split(config->http_proxy, &proxy_protocol, &proxy_username, &proxy_password, &proxy_host, &proxy_port);
```
this line exceeds 90 chars: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#line-length

erain replied:
Fixed.

edsiper commented on src/flb_utils.c:
```diff
@@ -870,3 +870,96 @@ int flb_utils_url_split(const char *in_url, char **out_protocol,
+ * Note: currently only HTTP is supported.
+ */
+int flb_utils_proxy_url_split(const char *in_url, char **out_protocol,
+                              char **out_username, char **out_password, char **out_host, char **out_port)
```
line length: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#line-length

erain replied:
Fixed.

edsiper reviewed (changes requested):
Please review the changes requested (comments)

erain commented:
Thanks a lot @edsiper for the review!! I will address your comments in the coming few hours. 

And just FYI I've also added the documentation for this feature in https://github.com/fluent/fluent-bit-docs/pull/414 .

edsiper commented:
thanks @erain !, make sure to address the code conflicts listed by github (you can rebase your work on top of GIT master)

erain commented:
Hi @edsiper , I've addressed all your comments and rebased from `master` (so no more merge confict). Unfortunately I needed to do another force push somehow and sorry for that. 

The only unaddressed comment was explained by @liuyuan10 already. 

Could you help review for one more time? Thanks!

edsiper commented:
Thanks for the fixes and clarification, the code looks good! 

Before to merge, would you please adjust the proper commits ?, the following commit is touching different files and having a single commit, I know is a bit redundant but that makes our maintainability easier in the future when checking changes:

- misc: address PR comments mostly related to readability

one commit per interface...

erain commented:
Thanks a lot for review @edsiper !

I broke the last commit into three smaller ones as you suggested and did another force push :)

edsiper commented:
thanks @erain !
